src/node/mod.rs

```diff
@@ -94,761 +84,223 @@ impl Node {
-        match op_index {
-            idx if idx == self.transport_rx_idx => {
-                let event = self.transport_rx.recv()?;
-                self.handle_transport_event(event);
-            }
-            idx if idx == self.timer_rx_idx => {
-                let token = self.timer_rx.recv()?;
-                self.handle_timeout(token);
-            }
-            _idx => return Err(RecvError),
-        };
-
-        self.handle_messages();
-
-        Ok(())
-    }
-
-    /// Returns whether this node is running or has been terminated.
-    pub fn is_running(&self) -> bool {
-        !matches!(self.stage, Stage::Terminated)
+    /// Is this the genesis node or not
+    pub fn is_genesis(&self) -> bool {
+        self.is_genesis
     }
 
     /// Returns the `PublicId` of this node.
-    pub fn id(&self) -> &PublicId {
-        self.core.id()
+    pub async fn id(&self) -> PublicId {
+        *self.stage.lock().await.full_id().public_id()
     }
 
     /// The name of this node.
-    pub fn name(&self) -> &XorName {
-        self.id().name()
+    pub async fn name(&self) -> XorName {
+        *self.id().await.name()
     }
 
     /// Returns connection info of this node.
-    pub fn our_connection_info(&mut self) -> Result<SocketAddr> {
-        self.core.our_connection_info()
+    pub async fn our_connection_info(&self) -> Result<SocketAddr> {
+        self.stage.lock().await.our_connection_info()
     }
 
     /// Our `Prefix` once we are a part of the section.
-    pub fn our_prefix(&self) -> Option<&Prefix> {
-        if let Stage::Approved(stage) = &self.stage {
-            Some(stage.shared_state.our_prefix())
-        } else {
-            None
-        }
+    pub async fn our_prefix(&self) -> Option<Prefix> {
+        self.stage.lock().await.our_prefix().cloned()
     }
 
-    /// Finds out if the given XorName matches our prefix. Returns error if we don't have a prefix
-    /// because we haven't joined any section yet.
-    pub fn matches_our_prefix(&self, name: &XorName) -> Result<bool> {
-        if let Some(prefix) = self.our_prefix() {
+    /// Finds out if the given XorName matches our prefix. Returns error
+    /// if we don't have a prefix because we haven't joined any section yet.
+    pub async fn matches_our_prefix(&self, name: &XorName) -> Result<bool> {
+        if let Some(prefix) = self.our_prefix().await {
             Ok(prefix.matches(name))
         } else {
             Err(Error::InvalidState)
         }
     }
 
     /// Returns whether the node is Elder.
-    pub fn is_elder(&self) -> bool {
-        self.stage
-            .approved()
-            .map(|stage| stage.is_our_elder(self.core.id()))
-            .unwrap_or(false)
+    pub async fn is_elder(&self) -> bool {
+        let name = self.name().await;
+        match self.stage.lock().await.approved() {
+            None => false,
+            Some(stage) => stage.shared_state.sections.our().elders.contains_key(&name),
+        }
```

Comment:
nice! aside, those Ids/pub_ids are really confusing to me, i cannot still get my head around them.

Reply:
I hear you. We use `id` and `public_id` interchangeably, which is probably not ideal. This has been brought up before and I think it will be simplified soon.